BLUESPAWN-client/src/util/filesystem/FileSystem.cpp

```diff
@@ -0,0 +1,458 @@
+					return false;
+				}
+				if(SetFilePointer(dwFileSize - dwCopyOffset) == INVALID_SET_FILE_POINTER){
+					LOG_ERROR("Can't set file pointer to " << offset << " in file " << FilePath << ".");
+					return false;
+				}
+				if(!WriteFile(hFile, buffer, dwCopySize, nullptr, nullptr)){
+					LOG_ERROR("Unable to write to " << FilePath << " at offset " << dwFileSize - dwCopyOffset << " (Error " << GetLastError() << ")");
+					return false;
+				}
+			}
+		}
+		//Write value over file at specified offset
+		if(SetFilePointer(offset) == INVALID_SET_FILE_POINTER) {
+			LOG_ERROR("Can't set file pointer to " << offset << " in file " << FilePath << ".");
+			return false;
+		}
+
+		if(!WriteFile(hFile, value, length, nullptr, nullptr)) {
+			LOG_ERROR("Failed to write to " << FilePath << " at offset " << offset << " with error " << GetLastError());
+			return false;
+		}
+
+		if(truncate) {
+			if(!SetEndOfFile(hFile)) {
+				LOG_ERROR("Couldn't truncate file " << FilePath);
+				return false;
+			}
+		}
+		LOG_VERBOSE(1, "Successfule wrote to " << FilePath << "at offset" << offset);
+		return true;
+	}
+
+	bool File::Read(OUT LPVOID buffer, __in_opt const unsigned long amount, __in_opt const long offset, __out_opt PDWORD amountRead) const {
+		SCOPE_LOCK(SetFilePointer(0), ResetFilePointer);
+		LOG_VERBOSE(2, "Attempting to read " << amount << " bytes from " << FilePath << " at offset " << offset);
+		if(!FileExists) {
+			LOG_ERROR("Can't write to " << FilePath << ". File doesn't exist.");
+			SetLastError(ERROR_FILE_NOT_FOUND);
+			return false;
+		}
+
+		if(SetFilePointer(offset) == INVALID_SET_FILE_POINTER) {
+			LOG_ERROR("Can't set file pointer to " << offset << " in file " << FilePath << ".");
+			return false;
+		}
+		if(!ReadFile(hFile, buffer, amount, amountRead, NULL)) {
+			LOG_ERROR("Failed to read from " << FilePath << " at offset " << offset << " with error " << GetLastError());
+			return false;
+		}
+		LOG_VERBOSE(1, "Successfully wrote " << amount << " bytes to " << FilePath);
+		return true;
+	}
+
+	AllocationWrapper File::Read(__in_opt unsigned long amount, __in_opt long offset, __out_opt PDWORD amountRead) const {
+		if(amount == -1){
+			amount = GetFileSize();
+		}
+		AllocationWrapper memory = { HeapAlloc(GetProcessHeap(), HEAP_ZERO_MEMORY, amount + 1), amount + 1, AllocationWrapper::HEAP_ALLOC };
+		bool success = Read(memory, amount, offset, amountRead);
+		return success ? memory : AllocationWrapper{ nullptr, 0 };
+	}
+
+	std::optional<std::string> File::GetMD5Hash() const {
+		LOG_VERBOSE(3, "Attempting to get MD5 hash of " << FilePath);
+		if(!FileExists) {
+			LOG_ERROR("Can't get MD5 hash of " << FilePath << ". File doesn't exist");
+			return std::nullopt;
+		}
+		//Function from Microsoft
+		//https://docs.microsoft.com/en-us/windows/desktop/SecCrypto/example-c-program--creating-an-md-5-hash-from-file-content
+		DWORD dwStatus = 0;
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Variable 'dwStatus' is assigned a value that is never used.](https://app.codacy.com/manual/ION28/BLUESPAWN/pullRequest?prid=4994516)